apps/blockchain/lib/blockchain/transaction.ex

```diff
@@ -190,8 +190,23 @@ defmodule Blockchain.Transaction do
   def execute(state, tx, block_header, config \\ EVM.Configuration.Frontier.new()) do
     {:ok, sender} = Transaction.Signature.sender(tx)
 
-    state_0 = begin_transaction(state, sender, tx)
+    {updated_state, remaining_gas, sub_state} =
+      state
+      |> begin_transaction(sender, tx)
+      |> apply_transaction(tx, block_header, sender, config)
 
+    {expended_gas, refund} = calculate_gas_usage(tx, remaining_gas, sub_state)
+
+    final_state =
+      updated_state
+      |> pay_and_refund_gas(sender, tx, refund, block_header)
+      |> clean_up_accounts_marked_for_destruction(sub_state, block_header)
+
+    # { σ', Υ^g, Υ^l }, as defined in Eq.(79) and Eq.(80)
+    {final_state, expended_gas, sub_state.logs}
+  end
+
+  defp apply_transaction(state, tx, block_header, sender, config) do
```

Comment:
typespec would be nice

Reply:
Added some in https://github.com/poanetwork/mana/pull/311/commits/08f862ab9eae240e29ff45fe1f41542c572fb984